man/man1/attr.1
.TH ATTR 1 "Extended Attributes" "Dec 2001" "XFS Compatibility API"
.SH NAME
attr \- extended attributes on XFS filesystem objects
.SH SYNOPSIS
.nf
\f3attr\f1 [ \f3\-LRSq\f1 ] \f3\-s attrname\f1 [ \f3\-V attrvalue\f1 ] \c
\f3pathname\f1
.sp .8v
\f3attr\f1 [ \f3\-LRSq\f1 ] \f3\-g attrname pathname\f1
.sp .8v
\f3attr\f1 [ \f3\-LRSq\f1 ] \f3\-r attrname pathname\f1
.sp .8v
\f3attr\f1 [ \f3\-LRSq\f1 ] \f3\-l pathname\f1
.sp .8v
.fi
.SH OVERVIEW
Extended attributes implement the ability for a user to attach
name:value pairs to objects within the XFS filesystem.
.P
They could be used to store meta-information about the file.
For example "character-set=kanji" could tell a document browser to
use the Kanji character set when displaying that document
and "thumbnail=..." could provide a reduced resolution overview of a
high resolution graphic image.
.PP
This document describes the
.I attr
command, which is mostly compatible with the IRIX command of the same name.
It is thus aimed specifically at users of the XFS filesystem - for
filesystem independent extended attribute manipulation, consult the
.IR getfattr (1)
and 
.IR setfattr (1)
documentation.
.P
In the XFS filesystem, the
.I names
can be up to 256 bytes in length, terminated by the first 0 byte.
The intent is that they be printable ASCII (or other character set)
names for the attribute.
The
.I values
can be up to 64KB of arbitrary binary data.
.P
Attributes can be attached to all types of XFS inodes:
regular files, directories, symbolic links, device nodes, etc.
.P
XFS uses 2 disjoint attribute name spaces associated with every
filesystem object.
They are the
.B root
and
.B user
address spaces.
The
.B root
address space is accessable only to the superuser,
and then only by specifying a flag argument to the function call.
Other users will not see or be able to modify attributes in the
.B root
address space.
The
.B user
address space is protected by the normal file permissions mechanism,
so the owner of the file can decide who is able to see and/or modify
the value of attributes on any particular file.
.SH DESCRIPTION
The
.I attr
utility allows the manipulation of extended attributes associated with
filesystem objects from within shell scripts.
.PP
There are four main operations that
.I attr
can perform:
.TP
.B GET
The
.B \-g attrname
option tells
.I attr
to search the named object and print (to \f4stdout\fP) the value
associated with that attribute name.
With the
.B \-q
flag, \f4stdout\fP will be exactly and only the value of the attribute,
suitable for storage directly into a file or processing via a piped command.
.TP
.B LIST
The
.B \-l
option tells
.I attr
to list the names of all the attributes that are associated with the object,
and the number of bytes in the value of each of those attributes.
With the
.B \-q
flag, \f4stdout\fP will be a simple list of only the attribute names,
one per line, suitable for input into a script.
.TP
.B REMOVE
The
.B \-r attrname
option tells
.I attr
to remove an attribute with the given name from the object if the
attribute exists.
There is no output on sucessful completion.
.TP
.B SET/CREATE
The
.B \-s attrname
option tells
.I attr
to set the named attribute of the object to the value read from \f4stdin\fP.
If an attribute with that name already exists,
its value will be replaced with this one.
If an attribute with that name does not already exist,
one will be created with this value.
With the
.B \-V attrvalue
flag, the attribute will be set to have a value of
.B attrvalue
and \f4stdin\fP will not be read.
With the
.B \-q
flag, \f4stdout\fP will not be used.
Without the
.B \-q
flag, a message showing the attribute name and the entire value
will be printed.
.PP
When the
.B \-L
option is given and the named object is a symbolic link,
operate on the attributes of the object referenced by the symbolic link.
Without this option, operate on the attributes of the symbolic link itself.
.PP
When the
.B \-R
option is given and the process has appropriate privileges,
operate in the
.I root
attribute namespace rather that the
.I USER
attribute namespace.
.PP
The
.B \-S
option is similar, except it specifies use of the
.I security
attribute namespace.
.PP
When the
.B \-q
option is given
.I attr
will try to keep quiet.
It will output error messages (to \f4stderr\fP)
but will not print status messages (to \f4stdout\fP).
.SH "NOTES"
The standard file interchange/archive programs
.IR tar (1),
and
.IR cpio (1)
will not archive or restore extended attributes,
while the
.IR xfsdump (8)
program will.
.SH "CAVEATS"
The list option present in the IRIX version of this command is not supported.
.I getfattr
provides a mechanism to retrieve all of the attribute names.
.SH "SEE ALSO"
getfattr(1),
setfattr(1),
attr_get(3),
attr_set(3),
attr_multi(3),
attr_remove(3),
attr(5),
and
xfsdump(8).


doc/CHANGES
attr-2.4.28 (13 January 2006)
	- Implement the IRIX list_attr interfaces in libattr.so,
	  the final piece of the IRIX attr compatibility API.
	- Put back the -L (list) option in attr(1).

attr-2.4.27 (05 December 2005)
	- Revert xattr.h/attributes.h to stating "Lesser GPL",
	  accidentally marked "GPL" in previous version.

attr-2.4.26 (10 November 2005)
	- Sync up build system (m4 macros, etc) with other projects
	- Update SGI copyright/licence notices

attr-2.4.25 (11 October 2005)
	- Add French translation from the debian-l10n-french folks
	  (thanks to Guilhelm Panaget)

attr-2.4.24 (10 August 2005)
	- Fix botched error reporting in attr_copy_file.c
	- Updated polish translation (both thanks to Jakub Bogusz)

attr-2.4.23 (3 June 2005)
	- attr_copy_{fd,file}: Reduce verboseness for ENOTSUP errors.
	  ENOTSUPP is common when copying between xattr enabled and
	  xattr disabled file systems.

attr-2.4.22 (21 February 2005)
	- Allocate the line buffer dynamically when reading from a file.
	  This mainly fixes restoring of large attributes.

attr-2.4.21 (31 January 2005)
	- Replace use of _POSIX_PATH_MAX with the larger PATH_MAX
	  (thanks to Andree Leidenfrost).

attr-2.4.20 (30 November 2004)
	- Licensing fixes and clarifications.
	- Update outdated email addresses.

attr-2.4.19 (29 September 2004)
	- Update m4 macros, incorporating some portability changes.

attr-2.4.18 (21 September 2004)
	- Add Dutch translation (thanks to Luk Claes)

attr-2.4.17 (8 September 2004)
	- Clarify setfattr usage information.
	- Fix email address in manual pages.

attr-2.4.16 (21 April 2004)
	- Updated Debian packaging dependency information.

attr-2.4.15 (02 February 2004)
	- Included Polish message translations by Jakub Bogusz.

attr-2.4.14 (14 January 2004)
	- When attr_copy_file is used on a symlink, process the
	  symlink instead of the file the symlink points to.
	- Improve wording in attr(5) man page.
	- Add the security namespace into the attr(1) command and
	  libattr IRIX-compatibility code for XFS commands to use.

attr-2.4.13 (06 January 2004)
	- Fix a permission bug in directory tree walking code.

attr-2.4.12 (13 October 2003)
	- Fix build on the GNU/Hurd platform for some Debian folks.

attr-2.4.11 (07 October 2003)
	- Man page tweaks

attr-2.4.10 (29 August 2003)
	- Fix a bug in nftw worker routines' error handling code.

attr-2.4.9 (27 August 2003)
	- Fix issues in the libmisc quote routine (from Ben Escoto).

attr-2.4.8 (04 August 2003)
	- Make quote/unquote return NULL if passed NULL.

attr-2.4.7 (29 July 2003)
	- Make quote return NULL if out of memory instead of exiting.

attr-2.4.6 (21 July 2003)
	- Add internal library for misc. functions: quote, unquote,
	  high_water_alloc.
	- Quote special characters in path names and attribute names.
	- Use high_water_alloc everywhere instead of malloc/realloc.
	- Update the test scripts.

attr-2.4.5 (03 July 2003)
	- Tweak the build so we don't always run msgmerge,
	  fix from Steve Langasek originally.

attr-2.4.4 (04 June 2003)
	- Some Debian-specific packaging updates.

attr-2.4.3 (26 April 2003)
	- Rework configure.in to make use of shared macros.

attr-2.4.2 (14 April 2003)
	- Found a better way to handle the trusted/xfsroot namespace
	  transition, from a suggestion from Andreas.  COMPAT_XFSROOT
	  is no longer checked, we figure this out on-the-fly now.
	- Fix configure tests that used AC_PATH_PROG incorrectly.

attr-2.4.1 (27 March 2003)
	- Cleanups to the build process, in particular the automatic
	  rpm package generation Makefile and spec file template.
	- Makepkgs script can now propagate failures back from make.
	- Change default --prefix to /usr.

attr-2.4.0 (26 February 2003)
	- Add symbol level versioning to libattr. This improves link
	  time consistency checks. The library versions are also
	  honored by RPM, so an RPM package built against say,
	  libattr.so.1(ATTR_1.1) won't link against
	  libattr.so.1(ATTR_1.0), which was not checked before.
	- Make the default check function used by attr_copy_{fd,file}()
	  accessible as attr_copy_check_permissions().
	- Increment the library version number, so that the attr_copy
	  functions can be checked for.
	- Fix some minor typos.

attr-2.3.0 (21 February 2003)
	- By default, we use the "trusted" attribute name prefix for
	  XFS ROOT attribute names, instead of the "xfsroot" prefix.
	  If the COMPAT_XFSROOT environment variable is set, however,
	  the old (now deprecated) names are used.  This is useful for
	  XFS kernel code older than, well, today.  Newer kernels will
	  work correctly with the new tools (but not the old tools).
	- Added routines to libattr for copying extended attributes -
	  attr_copy_file and attr_copy_fd.
	- libattr.so version updated to reflect these changes.
	- Update licensing notice in system call man pages for aeb.

attr-2.2.0 (30 November 2002)
	- Complete the internationalisation support, and added initial
	  German translation from Andreas.

attr-2.1.2 (22 November 2002)
	- Add a description of trusted extended attributes to attr(5).

attr-2.1.1 (31 October 2002)
	- Minor updates to the Debian packaging.

attr-2.1.0 (19 October 2002)
	- Change the devel .rpm and .deb package names so that they
	  now use a lib prefix.
	- Several packaging changes related to this to ensure upgrades
	  work and depenency information isn't lost.

attr-2.0.12 (07 October 2002)
	- Fix: the m68k architecture is identified by __mc68000__
	  rather than by __m68k__.
	- Correction in the attr(5) manual page.
	- Add test cases for empty EA values, and for a bad
	  ACL extended attribute name (there were bugs with these).

attr-2.0.11 (04 September 2002)
	- Make the test/run script return a proper status code.
	- Fix up the error message returned for ENODATA rather
	  than ENOATTR, so if ENOATR is assigned a separate number
	  we don't mess with ENOATTR's text message.

attr-2.0.10 (28 August 2002)
	- Add system call numbers for the m68k architecture

attr/attr.c
/*
 * Copyright (c) 2000-2002,2004 Silicon Graphics, Inc.
 * All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it would be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write the Free Software Foundation,
 * Inc.,  51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <sys/types.h>
#include <sys/param.h>
#include <sys/stat.h>
#include <stdio.h>
#include <fcntl.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>
#include <locale.h>

#include <attr/attributes.h>
#include "config.h"

#define	SETOP		1		/* do a SET operation */
#define	GETOP		2		/* do a GET operation */
#define	REMOVEOP	3		/* do a REMOVE operation */

static char *progname;

void
usage(void)
{
	fprintf(stderr, _(
"Usage: %s [-LRSq] -s attrname [-V attrvalue] pathname  # set value\n"
"       %s [-LRSq] -g attrname pathname                 # get value\n"
"       %s [-LRSq] -r attrname pathname                 # remove attr\n"
"      -s reads a value from stdin and -g writes a value to stdout\n"),
		progname, progname, progname);
	exit(1);
}

int
main(int argc, char **argv)
{
	char *attrname, *attrvalue, *filename;
	int attrlength;
	int opflag, ch, error, follow, verbose, rootflag, secureflag;

	progname = basename(argv[0]);

	setlocale(LC_CTYPE, "");
	setlocale(LC_MESSAGES, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);

	/*
	 * Pick up and validate the arguments.
	 */
	verbose = 1;
	follow = opflag = rootflag = secureflag = 0;
	attrname = attrvalue = NULL;
	while ((ch = getopt(argc, argv, "s:V:g:r:qLRS")) != EOF) {
		switch (ch) {
		case 's':
			if ((opflag != 0) && (opflag != SETOP)) {
				fprintf(stderr,
				    _("Only one of -s, -g, or -r allowed\n"));
				usage();
			}
			opflag = SETOP;
			attrname = optarg;
			break;
		case 'V':
			if ((opflag != 0) && (opflag != SETOP)) {
				fprintf(stderr,
				    _("-V only allowed with -s\n"));
				usage();
			}
			opflag = SETOP;
			attrvalue = optarg;
			break;
		case 'g':
			if (opflag) {
				fprintf(stderr,
				    _("Only one of -s, -g, or -r allowed\n"));
				usage();
			}
			opflag = GETOP;
			attrname = optarg;
			break;
		case 'r':
			if (opflag) {
				fprintf(stderr,
				    _("Only one of -s, -g, or -r allowed\n"));
				usage();
			}
			opflag = REMOVEOP;
			attrname = optarg;
			break;
		case 'L':
			follow++;
			break;
		case 'R':
			rootflag++;
			break;
		case 'S':
			secureflag++;
			break;
		case 'q':
			verbose = 0;
			break;
		default:
			fprintf(stderr,
			    _("Unrecognized option: %c\n"), (char)ch);
			usage();
			break;
		}
	}
	if (optind != argc-1) {
		fprintf(stderr, _("A filename to operate on is required\n"));
		usage();
	}
	filename = argv[optind];

	/*
	 * Break out into option-specific processing.
	 */
	switch (opflag) {
	case SETOP:
		if (attrvalue == NULL) {
			attrvalue = malloc(ATTR_MAX_VALUELEN);
			if (attrvalue == NULL) {
				perror("malloc");
				exit(1);
			}
			attrlength =
				fread(attrvalue, 1, ATTR_MAX_VALUELEN, stdin);
		} else {
			attrlength = strlen(attrvalue);
		}
		error = attr_set(filename, attrname, attrvalue,
					   attrlength,
					   (!follow ? ATTR_DONTFOLLOW : 0) |
					   (secureflag ? ATTR_SECURE : 0) |
					   (rootflag ? ATTR_ROOT : 0));
		if (error) {
			perror("attr_set");
			fprintf(stderr, _("Could not set \"%s\" for %s\n"),
					attrname, filename);
			exit(1);
		}
		if (verbose) {
			printf(_("Attribute \"%s\" set to a %d byte value "
			       "for %s:\n"), attrname, attrlength, filename);
			fwrite(attrvalue, 1, attrlength, stdout);
			printf("\n");
		}
		break;

	case GETOP:
		attrvalue = malloc(ATTR_MAX_VALUELEN);
		if (attrvalue == NULL) {
			perror("malloc");
			exit(1);
		}
		attrlength = ATTR_MAX_VALUELEN;
		error = attr_get(filename, attrname, attrvalue,
					   &attrlength,
					   (!follow ? ATTR_DONTFOLLOW : 0) |
					   (secureflag ? ATTR_SECURE : 0) |
					   (rootflag ? ATTR_ROOT : 0));
		if (error) {
			perror("attr_get");
			fprintf(stderr, _("Could not get \"%s\" for %s\n"),
					attrname, filename);
			exit(1);
		}
		if (verbose) {
			printf(_("Attribute \"%s\" had a %d byte value "
				"for %s:\n"), attrname, attrlength, filename);
		}
		fwrite(attrvalue, 1, attrlength, stdout);
		if (verbose) {
			printf("\n");
		}
		break;

	case REMOVEOP:
		error = attr_remove(filename, attrname,
					      (!follow ? ATTR_DONTFOLLOW : 0) |
					      (secureflag ? ATTR_SECURE : 0) |
					      (rootflag ? ATTR_ROOT : 0));
		if (error) {
			perror("attr_remove");
			fprintf(stderr, _("Could not remove \"%s\" for %s\n"),
					attrname, filename);
			exit(1);
		}
		break;

	default:
		fprintf(stderr,
			_("At least one of -s, -g, or -r is required\n"));
		usage();
		break;
	}

	return(0);
}
